src/spox/_attributes.py

```diff
@@ -50,20 +46,44 @@ def value(self) -> T:
         return self._value
 
     def _validate(self):
-        if self._to_onnx("dummy").type != self._attribute_proto_type_int:
-            raise TypeError(
-                f"Unable to instantiate `{type(self).__name__}` with value of type `{type(self.value).__name__}`."
-            )
+        try:
+            type_in_onnx = self._to_onnx_deref("dummy").type
+        except TypeError as e:
+            # TypeError: 'a' has type str, but expected one of: int -- this is raised from within protobuf
+            # when .extend()-ing with the wrong type.
+            if "but expected one of" in str(e):
+                raise self._get_pretty_type_exception() from e
+            else:
+                raise
+
+        if type_in_onnx != self._attribute_proto_type_int:
+            raise self._get_pretty_type_exception()
 
     def _to_onnx(self, key: str) -> AttributeProto:
         if isinstance(self._value, _Ref):
             return self._value._to_onnx(key)
         return self._to_onnx_deref(key)
 
+    @property
+    @abc.abstractmethod
+    def _attribute_proto_type_int(self) -> int:
+        raise NotImplementedError
+
+    @abc.abstractmethod
     def _to_onnx_deref(self, key: str) -> AttributeProto:
         """Conversion method for the dereferenced case."""
         raise NotImplementedError
 
+    def _get_pretty_type_exception(self):
+        if isinstance(self.value, tuple) and len(self.value):
+            tuple_types = ", ".join({type(v).__name__ for v in self.value})
+            value_type = f"tuple[{tuple_types}, ...]"
+        else:
+            value_type = type(self.value).__name__
+        return AttributeTypeError(
+            f"Unable to instantiate `{type(self).__name__}` with value of type `{value_type}`."
+        )
```

Comment:
I think we can keep it simple at this point. If we tell the user that it was not possible to instantiate ... from the provided values then they probably know what's up. Furthermore, in many cases the onnx-provided error is quite good and we raise it as well. E.g.:
```python
make_attribute("foo", ["bar", ], attr_type=AttributeProto.FLOATS)
```
> TypeError: 'bar' has type str, but expected one of: int, float

Reply:
It does provide some more context, while also keeping the given one, e.g. the concrete attr class and all the types in the tuple. It also makes the errors more consistent.